### PR TITLE
drivers: serial: Add error code to uart_rx_buf_rsp (unstable API change)

### PR DESCRIPTION
**Pull request by nordic-krch**
API did not cover a case when uart_rx_buf_rsp is called when receiver
is already disabled. It may happen if uart_rx_buf_rsp is called too
late and active transfer is already finished. In that case -EAGAIN is
returned.

Added implementation for nrf_uarte.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

### CODE REVIEW

**Comment by nordic-krch** [comment]
DNM until other implementations are updated.

**Review by carlescufi — APPROVED** [review]
This looks reasonable to me.

**Review by anangl — COMMENTED** [review]
Looks good, except for the error code used. I think we should find some more suitable.

File: include/drivers/uart.h [hunk 1]
```diff
@@ -560,6 +560,7 @@ static inline int z_impl_uart_rx_enable(struct device *dev, uint8_t *buf,
  *
  * @retval -ENOTSUP If not supported.
  * @retval -EBUSY   Next buffer already set.
+ * @retval -EAGAIN  Receiver is already disabled (function called too late?).
```
[the remark this answers is not in the document]
Comment (de-nordic) [reply include/drivers/uart.h:563]: What is returned from other devices in such case?

[the remark this answers is not in the document]
Comment (de-nordic) [reply include/drivers/uart.h:563]: If we do not distinguish between disabled and non-existent devices, `NODEV` could be used.or use some of operational errors `ENOTCONN` or `ESHUTDOWN`, which is ` Cannot send after transport endpoint shutdown.` according to linux man page for errno, and quite fits `Receiver is already disabled (function called too late?).`

[the remark this answers is not in the document]
Comment (nordic-krch) [reply include/drivers/uart.h:563]: `ESHUTDOWN` sounds ok to me, @anangl what do you think?

[the remark this answers is not in the document]
Comment (nordic-krch) [reply include/drivers/uart.h]: turns out that 2/3 implementations already were handling that case and returned `-EACCES`, decided to keep it that way as some may already use it.

**Comment by nordic-krch** [comment]
Removed DNM. Also added a commit for nrf_uart where i added protection against race condition between disabling RX and calling `uart_rx_buf_rsp`.

**Review by anangl — CHANGES_REQUESTED** [review]

**Review by anangl — APPROVED** [review]

**Review by de-nordic — APPROVED** [review]

**Pull request merged**